exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/DrillFileSystemCache.java

```diff
@@ -0,0 +1,99 @@
+ *   {@link org.apache.hadoop.fs.FileSystem} cache has the following bad behavior:
+ * </p>
+ * <ul>
+ *   <li>
+ *     The {@link org.apache.hadoop.conf.Configuration} object is not considered when constructing keys for the {@link org.apache.hadoop.fs.FileSystem} cache of
+ *     {@link org.apache.hadoop.fs.FileSystem} objects.
+ *   </li>
+ *   <li>
+ *      The {@link org.apache.hadoop.fs.FileSystem} cache does not honor the <b>fs.default.name</b> property when constructing keys, only <b>fs.defaultFS</b> is used to construct
+ *      keys in the cache.
+ *   </li>
+ * </ul>
+ *
+ * <h4>Usage</h4>
+ *
+ * <ul>
+ *   <li>
+ *     A prerequisite for usage is that all {@link org.apache.hadoop.conf.Configuration} objects are normalized with
+ *     {@link org.apache.drill.exec.store.dfs.DrillFileSystem#normalize(Configuration)}.
+ *   </li>
+ *   <li>
+ *     This cache should only be used from {@link org.apache.drill.exec.store.dfs.DrillFileSystem}.
+ *   </li>
+ * </ul>
+ *
+ * <h4>TODO</h4>
+ *
+ * <ul>
+ *   <li>
+ *     Drill currently keeps a {@link org.apache.drill.exec.store.dfs.DrillFileSystem} open indefinitely. This will be corrected
+ *     in DRILL-6608. As a result this cache currently has no methods to remove {@link org.apache.hadoop.fs.FileSystem} objects
+ *     after they are created.
+ *   </li>
+ * </ul>
+ */
+class DrillFileSystemCache {
+  private Map<Map<String, String>, FileSystem> cache = new HashMap<>();
+
+  /**
+   * If a {@link org.apache.hadoop.fs.FileSystem} object corresponding to the given {@link org.apache.hadoop.conf.Configuration}
+   * exists in the cache, then it is returned. If no corresponding {@link org.apache.hadoop.fs.FileSystem} exist, then it is created,
+   * added to the cache, and returned.
+   * @param configuration The {@link org.apache.hadoop.conf.Configuration} corresponding to the desired {@link org.apache.hadoop.fs.FileSystem}
+   *                      object. It is expected that this configuration is first normalized with {@link org.apache.drill.exec.store.dfs.DrillFileSystem#normalize(Configuration)}.
+   * @return The {@link org.apache.hadoop.fs.FileSystem} object corresponding to he given {@link org.apache.hadoop.conf.Configuration}.
+   * @throws IOException An error when creating the desired {@link org.apache.hadoop.fs.FileSystem} object.
+   */
+  protected synchronized FileSystem get(final Configuration configuration) throws IOException {
+    final Map<String, String> map = new HashMap<>(configToMap(configuration));
+
+    if (!cache.containsKey(map)) {
+      cache.put(map, FileSystem.newInstance(configuration));
+    }
+
+    return cache.get(map);
+  }
+
+  static Map<String, String> configToMap(final Configuration configuration) {
+    Preconditions.checkNotNull(configuration);
+    final Map<String, String> map = new HashMap<>();
+
+    for (Map.Entry<String, String> entry: configuration) {
+      map.put(entry.getKey().trim(), entry.getValue());
```

Comment:
In this implementation we are only considering `Configuration` object in Cache key. How about other parameters which [HadoopFileSystem cache](https://github.com/apache/hadoop/blob/branch-2.7.1/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java#L2812) is using ? Specially I am concerned about the UGI object from which the user information is retrieved which is later checked for access control.

Also for handling `fs.default.name` and `fs.defaultFS` once it's set in the configuration parameter and FileSystem instance is created then [FileSystemCache does use information from there based on UR](https://github.com/apache/hadoop/blob/branch-2.7.1/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java#L2806)I (which is created using those parameters).

Reply:
- The UGI object is only created and used in the Key when FileSystem.getUnique is called. The call DrillFileSystem was using previously was FileSystem.get, so the cache never used UGI objects as part of the key for our case.  The behavior is pre-existing and I'm not sure if it's a bug or now. We should determine if it is a bug or not, if it is I would like to fix that as a separate Jira. What are your thoughts?

- For your second point, you are right. However, there is a bug in the logic of FileSystem.getDefaultUri (which is used to construct the key) that if both fs.defaultFS and fs.default.name are defined in a configuration the deprecated option (fs.defaultFS) wins. Also it does not take into account the rest of the configuration object in the key. So if you require two different configurations for the same file system you are out of luck, only the first one wins.

Reply:
- Usageof UGI object is done in such a way that FileSystem instance is created in doAs block of it. UGI holds the information about the trusted user. Now when DrillFilesystem constructor is called it calls `FileSystem.get`. This internally calls `CACHE.get(uri, conf)` which creates an instance of `CACHE.Key` (See [here](https://github.com/apache/hadoop/blob/branch-2.7.1/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java#L371)). In the `Key` constructor the call to `UGI.getCurrentUser()` returns the context of user under which the application called the constructor. Later this user context is also used in `initialize` of specific implementation of FileClient. Like for HDFS client it is used [here](https://github.com/apache/hadoop/blob/branch-2.7.1/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DFSClient.java#L647) again. But for future call if DrillFileSystem is called under same user block then the cache version of correct filesystem object will be returned. So in short we have to use it otherwise there will be security regression.

- I don't understand why are we preferring deprecated property over the new one ? `fs.default.name` is deprecate one and `fs.defaultFS` is new property. (See [here](https://hadoop.apache.org/docs/r2.7.2/hadoop-project-dist/hadoop-common/DeprecatedProperties.html)) . Also [Configuration is a static field of FileSystem class](https://github.com/apache/hadoop/blob/branch-2.7.1/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java#L3301) so for same scheme there cannot be 2 different configuration instances. Looks like that was the intention and hence CACHE also doesn't use it.

Reply:
- The code for CACHE.get calls the Key(uri, conf) constructor https://github.com/apache/hadoop/blob/branch-2.7.1/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java#L2667 . That constructor does not initialize the ugi object, so it is not used as part of the key. https://github.com/apache/hadoop/blob/branch-2.7.1/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java#L2801 
The UGI object is only used when CACHE.getUnique is called. In that method the Key(uri, conf, unique) constructor is called https://github.com/apache/hadoop/blob/branch-2.7.1/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java#L2673 . Only in that case does the Key constructor call UGI.getCurrentUser() . https://github.com/apache/hadoop/blob/branch-2.7.1/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java#L2801 . So if not using the UGI object is a security issue, it is a pre-existing issue. Please double check the code I referenced to make sure my analysis is correct.
 - You are right I had the deprecated and non-deprecated property names mixed up. Thank you very much for catching this! 
 - While there is static conf field in the FileSystem class, no one seems to use it. For example DistrubutedFileSystem calls the setConf method https://github.com/apache/hadoop/blob/branch-2.7.1/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/DistributedFileSystem.java#L137, which sets a conf variable only for that instance of the class https://github.com/apache/hadoop/blob/branch-2.7.1/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/conf/Configured.java (DistributedFileSystem extends FileSystem which extends Configured). In fact as far as I can tell from my IDE the static conf variable is never used. Whenever the FileSystem class uses the configuration it calls getConf() which is a method of Configured and which returns the configuration that is stored on a per instance basis. So I think the configuration should still be used in the key for caching.

Reply:
Makes sense and probably that's the reason `Cache` only uses `Scheme` and `authority` properties which are derived from `fs.defaultFS` parameter. I guess the proxy servers are provided as `authority` strings. All Kerberos and security related parameters are encapsulated within `UGI` object.